### ISSUE TITLE
Sporadical lint/typecheck error in monorepo

### ISSUE THREAD

**Issue by anotheri** [oxc-project/tsgolint#987]
In quite large monorepo (50+ packages, most of them are relatively small ones though), running oxlint --type-aware (per package) and typecheck (tsgo, also per package) causes randomly (in ~50% chance) false positive errors on lint.

🔴 this behaviour happens (almost every 2nd run) with: 
- `oxlint-tsgolint: 0.23.0` 
- `oxlint: 1.66.0` (or newer) 
- `@typescript/native-preview: 7.0.0-dev.20260521.1` or (newer), 
- and fwiw in this case it happens even when i run only lint task over all the packages (it executes them in parallel processes, with some concurrency based on cpus, i guess)

🟡  it doesn't happen (tbh, it happens but quite rarely, ~10% chance of error in typecheck, not even in lint task) with:
- `oxlint-tsgolint: 0.22.1` 
- and the latest `oxlint` and `@typescript/native-preview`

🟢 also it doesn't happen (never!) with:
- `oxlint-tsgolint: 0.23.0`
- `@typescript/native-preview: 7.0.0-dev.20260513.1`
- `oxlint: 1.65.0`
- but as soon as i update oxlint or @typescript/native-preview here the issue happens again.

Could you please help me to figure that out? Thanks in advance.

**Comment by camc314** [comment]
This might be the same as #21752

Can you clarify what exactly the false positives are? Are they lint errors similar to #21752 ? or type errors?

Who is producing the type errors? is it us (oxlint-tsgolint), or `@typescript-/native-preview`?

thanks!

**Comment by anotheri** [comment]
@camc314 

fwiw i run tasks via `"@moonrepo/cli": "^2.2.6"` now (and before i've used `"@moonrepo/cli": "^2.2.4"` but result more or less the same, so probably not important) on the following environment:

```
 System:
    OS: macOS 26.5
    CPU: (16) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 4.65 GB / 64.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 24.16.0 - /Users/anotheri/.proto/tools/node/24.16.0/bin/node
    Yarn: 1.22.22 - /usr/local/bin/yarn
    npm: 11.13.0 - /Users/anotheri/.proto/tools/node/24.16.0/bin/npm
    pnpm: 11.3.0 - /Users/anotheri/.proto/shims/pnpm
    bun: 1.1.34 - /Users/anotheri/.proto/shims/bun
    Watchman: 2026.05.25.00 - /usr/local/bin/watchman
```

tasks:
- `lint` - alias for `oxlint --type-aware --max-warnings=0`
-  `typecheck` - alias for `tsgo`

so i've just tried it with:
- "@typescript/native-preview": "7.0.0-dev.20260527.2",
- "oxlint": "^1.67.0",
- "oxlint-tsgolint": "^0.23.0",

and i've tested it by running those tasks sequentially 10 times in a raw (via python script) through all the packages (50+) in parallel and here is summary:

## Benchmark Results: 10 runs per task

### Pass/Fail Summary

| Task | Pass | Fail | Failure Rate |
|---|---|---|---|
| `moon run :lint -f` | 9 | **1** | 10% |
| `moon run :typecheck -f` | 4 | **6** | 60% |
| `moon run :lint :typecheck -f` | 4 | **6** | 60% |

### Root Cause Analysis

The failures are **not random code errors** — they fall into two distinct categories:

#### 1. Intermittent `tsgo` module resolution failures (12 of 12 failures)

`tsgo` intermittently fails to resolve `node_modules` type declarations. The same packages fail differently on each run — **the failing package is random**, but the error pattern is always the same:

| Error | Count | Packages Affected |
|---|---|---|
| `TS2307: Cannot find module 'vitest'` | 32 | sky-liquidity, sky-button, sky-number, sky-datetime, theme |
| `TS2882: Cannot find module 'vitest-canvas-mock'` | 3 | data-roles, data-users, data-bot-commands |
| `TS2339: Property 'toBeInTheDocument'/'toHaveAttribute' on Assertion` | 6 | sky-logo-brand |
| `TS2882: Cannot find module '@testing-library/jest-dom/vitest'` | 1 | sky-logo-brand |
| `TS2307: Cannot find module '@testing-library/react'` | 1 | theme |
| `TS7006: Parameter implicitly has 'any' type` | 3 | sky-number |

The `TS2307` errors (cannot find module) are the dominant failure — they indicate `tsgo` is intermittently failing to read `node_modules` or resolve package types. This is likely a **race condition or filesystem cache issue** with `tsgo`.

#### 2. Process crashes (2 of 12 failures)

- **SIGPIPE** in `data-auth-api:lint` — oxlint process died mid-pipe
- **SIGSEGV** (segfault) in `data-signup:typecheck` — tsgo crashed

### Top 11 Error Summary (across all 20 typecheck runs)

| # | Error | Count |
|---|---|---|
| 1 | `TS2307: Cannot find module 'vitest'` | 32 |
| 2 | `typescript(no-unsafe-call): Unsafe call of error type` | 12 |
| 3 | `TS2339: Property 'toHaveAttribute' on Assertion` | 4 |
| 4 | `TS7006: Parameter implicitly has 'any' type` | 3 |
| 5 | `TS2882: Cannot find module 'vitest-canvas-mock'` | 3 |
| 6 | `typescript(no-unsafe-member-access): Unsafe member access .toBe` | 3 |
| 7 | `TS2339: Property 'toBeInTheDocument' on Assertion` | 2 |
| 8 | `typescript(no-unsafe-member-access): Unsafe member access .toBeNull` | 1 |
| 9 | `TS2882: Cannot find module '@testing-library/jest-dom/vitest'` | 1 |
| 10 | `TS2307: Cannot find module '@testing-library/react'` | 1 |
| 11 | `typescript(no-unsafe-assignment): Unsafe assignment of an error typed value.` | 1 |

### Key Findings

1. **`:lint` is mostly stable** — 9/10 passes
2. **`:typecheck` is flaky 60% of the time** — root cause is `tsgo` intermittently failing to resolve `node_modules` type declarations
3. **The specific package that fails changes every run** — this is a systemic `tsgo` issue, not a per-package issue
4. **Combined `:lint :typecheck` inherits the typecheck flakiness** — same 60% failure rate


## PS

And it looks more like i have to open this issue in `tsgo` or `vitest` (lint seems the least issue here), but i'll try to run the same benchmark a couple more times on less recent and more stable (at least for me) dependencies versions, and let you know.

**Comment by anotheri** [comment]
also fwiw every time it fails i see 5 `tsgolint` processes in activity monitor (it seems 5 is default concurrency on moon runner), and cpu (not the most efficient, yep) goes quite high as well (for some reasons it doesn't show cpu per process though):

<img width="1449" height="1088" alt="Image" src="https://github.com/user-attachments/assets/6ca5c89f-0b5c-4c35-bb7c-e7451e942add" />

**Comment by anotheri** [comment]
so, i did more tests mainly for `typecheck` task using the following dependencies:
- `oxlint-tsgolint: 0.23.0`
- `oxlint: 1.67.0`
- `@typescript/native-preview` between `7.0.0-dev.20260513.1` and `7.0.0-dev.20260527.2` 
 
and it looks like `7.0.0-dev.20260515.1` release is where regression is introduced (8/10 passed), so the last stable version for me is `7.0.0-dev.20260514.1`.

most probably related to commit from May 14: 
> Speed up realpath on Linux and macOS (#3838) (https://github.com/microsoft/typescript-go/commit/a8b2a52b76dd41d4612dc404ae4449eb9185d387) 

This changes realpath behavior which is directly related to symlink/module resolution through pnpm (which i use here).

---

## Root Cause Analysis

### Regression Version
**`@typescript/native-preview@7.0.0-dev.20260515.1`** (released May 15, 2026)

### Binary Search Results
| Version | Result |
|---|---|
| 20260513.1 | 10/10 pass |
| 20260514.1 | 10/10 pass |
| **20260515.1** | **2/10 fail** (regression introduced) |
| 20260516.1 | 3/10 fail |
| 20260520.1 | 4/10 fail |
| 20260527.2 | 5/10 fail |

### Root Cause Commit
**[Speed up realpath on Linux and macOS (#3838)](https://github.com/microsoft/typescript-go/commit/a8b2a52b76dd41d4612dc404ae4449eb9185d387)** — authored by jakebailey on May 14, 2026.

### What Changed
The commit replaced Go's `filepath.EvalSymlinks` (which does `lstat` per path component) with a faster O(1) syscall approach:

**macOS**: Uses `open(path, O_EVTONLY|O_NONBLOCK|O_CLOEXEC)` + `fcntl(fd, F_GETPATH, buf)` to resolve symlinks

**Linux**: Uses `O_PATH` + `readlink("/proc/self/fd/<fd>")` trick

### Why It Breaks pnpm Workspaces
pnpm creates deeply nested symlink chains:
```
packages/components/sky-liquidity/node_modules/vitest → 
  ../../../../node_modules/.pnpm/vitest@4.1.7_.../node_modules/vitest
```

The new `realpath` implementation using `O_EVTONLY` on macOS appears to intermittently fail to correctly resolve these symlink chains. The `O_EVTONLY` flag is designed for event monitoring and "doesn't require read permission" — this may cause subtle differences in how it follows symlinks compared to the original `filepath.EvalSymlinks` approach.

**Comment by anotheri** [comment]
## Benchmark Results:

with `@typescript/native-preview@7.0.0-dev.20260514.1` (stable)

| Task | Pass | Fail | Failure Rate |
|---|---|---|---|
| `moon run :lint -f` | 14 | 6 | 30% |
| `moon run :typecheck -f` | **20** | **0** | **0%** |
| `moon run :lint :typecheck -f` | ~10 | ~4 | ~29% |

**Key finding**: `typecheck` is **100% stable** with the pinned version. The `lint` failures are from `oxlint-tsgolint` which internally uses `tsgo` for type-aware linting — same underlying race condition but in a different context.

## Root Cause

The issue you found — **[microsoft/typescript-go#4060](https://github.com/microsoft/typescript-go/issues/4060)** — is exactly our problem:

- **Data race in module resolver**: `OrderedMap.Keys` is not thread-safe under concurrent goroutines
- **Affects pnpm monorepos** with parallel `tsgo` execution
- **Regression introduced** around `7.0.0-dev.20260515.1` (May 15, 2026)
- **Workaround**: `GOMAXPROCS=1` or `GOMAXPROCS=2` (reduces internal goroutine parallelism)

## Recommended Actions

1. **Pin** `@typescript/native-preview` to `7.0.0-dev.20260514.1` until upstream fix
2. **Monitor** issue #4060 for the official fix
3. **Alternative**: Set `GOMAXPROCS=2` in your moon task environment if you need newer tsgo features

**Comment by anotheri** [comment]
it's fixed with `@typescript/native-preview: 7.0.0-dev.20260603.1`
tested with both `oxlint: 1.65.0` and `oxlint: 1.68.0`, everything works as expected now

**Comment by anotheri** [comment]
@camc314 although... i test more with 1.68 and getting lint errors only (chance is low 20-30% i'd say, but still happens) ~when i run both tasks together~ even if i run only lint task over ~100 packages... so typecheck is fixed, but lint is not. Do tsgolint dependencies need to be updated with  `@typescript/native-preview: 7.0.0-dev.20260603.1` separately or it supposed to use my local ones?

**Comment by camc314** [comment]
yes - i'll need to pull in the typescript-go change & do a release.

This usually happens on Fridays

**Comment by anotheri** [comment]
ok, then i'll double check and close this issue after next release

**Comment by alvarosevilla95** [comment]
It looks like main now has a tsgo submodule with the [commit that fixes this issue](https://github.com/microsoft/typescript-go/pull/4178), but no release has been made that includes it yet. 

Would it be possible to publish a new version to resolve this?

**Comment by anotheri** [comment]
confirmed, it's fixed in `oxlint-tsgolint: 0.24.0`
